components/shared/onboarding/Onboarding.tsx

```diff
@@ -0,0 +1,320 @@
+import { updateUser } from "@/lib/actions/user.actions";
+import OnboardingFour from "./OnboardingFour";
+import CustomButton from "../CustomButton";
+import { Progress } from "@/components/ui/progress";
+import OnboardingTick from "./OnboardingTick";
+import UploadPhoto from "../profile/edit/UploadPhoto";
+import { useRouter } from "next/navigation";
+
+interface OnboardingProps {
+  children?: React.ReactNode;
+  step: string;
+  user: Partial<IUser>;
+}
+
+const Onboarding = ({ step, user }: OnboardingProps) => {
+  const [currentStep, setCurrentStep] = React.useState(step);
+  const router = useRouter();
+  const [pending, startTransition] = useTransition();
+  const {
+    handleSubmit,
+    control,
+    register,
+    getValues,
+    formState: { errors },
+  } = useForm<z.infer<typeof CompleteProfileEditSchema>>({
+    resolver: zodResolver(CompleteProfileEditSchema),
+    defaultValues: {
+      learningGoals: user.learningGoals || [],
+      experiences: user.experiences?.map((exp) => ({ name: exp })) || [],
+      availability: {
+        startTime: user?.startTime || new Date(),
+        endTime: user?.endTime || new Date(),
+      },
+      newProjects: user?.newProjects || false,
+      technologies: user?.technologies || [],
+      portfolio: user?.portfolio || "",
+      email: user?.email || "",
+    },
+  });
+  const {
+    fields: goalFields,
+    append: appendGoal,
+    remove: removeGoal,
+  } = useFieldArray<z.infer<typeof CompleteProfileEditSchema>>({
+    control,
+    name: "learningGoals",
+  });
+
+  const {
+    fields: experienceFields,
+    append: appendExperience,
+    remove: removeExperience,
+  } = useFieldArray<z.infer<typeof CompleteProfileEditSchema>>({
+    control,
+    name: "experiences",
+  });
+  const changeStep = async () => {
+    const nextStep = parseInt(currentStep) + 1;
+    const {
+      fullname,
+      portfolio,
+      availability,
+      learningGoals,
+      email,
+      experiences,
+      newProjects,
+    } = getValues();
+    try {
+      if (nextStep === 2) {
+        await updateUser({
+          updateData: {
+            fullname,
+            portfolio,
+            email,
+          },
+        });
+      }
+      if (nextStep === 3) {
+        await updateUser({
+          updateData: {
+            learningGoals,
+          },
+        });
+      }
+      if (nextStep === 4) {
+        await updateUser({
+          updateData: {
+            experiences: experiences?.map((experience) => experience.name),
+          },
+        });
+      }
+      if (nextStep === 5) {
+        startTransition(() => {
+          updateUser({
+            updateData: {
+              newProjects,
+              startTime: availability?.startTime,
+              endTime: availability?.endTime,
+            },
+          });
+        });
+      }
+    } catch (error) {
+      console.log(error);
+    }
+
+    setCurrentStep(nextStep.toString());
+  };
```

Comment:
Because we added a list of fields to that stepMap object- we can just select which fields we need to update here. This has a few benefits- shorter code. And it's expandable. If we need to change things we just adjust our map to render a different component and add the fields. 

```js
  const changeStep = async () => {
    const nextStep = parseInt(currentStep) + 1;

    const fields = stepMap[currentStep].fields;

    const filteredObject = Object.keys(getValues()).reduce((acc, key) => {
      if (fields.includes(key)) {
        // @ts-ignore
        acc[key] = getValues()[key];
      }
      return acc;
    }, {} as any);

    if (filteredObject.experiences) {
      filteredObject.experiences = filteredObject.experiences.map(
        (experience: any) => experience.name
      );
    }

    try {
      await updateUser({
        updateData: filteredObject,
      });
    } catch (error) {
      console.log(error);
    }

    setCurrentStep(nextStep.toString());
  };
  ```